### PR TITLE
bpo-33649: Add low-level APIs index.

### PR DESCRIPTION
**Pull request by 1st1**
https://bugs.python.org/issue33649

### CODE REVIEW

File: Doc/library/asyncio-api-index.rst [hunk 1]
```diff
@@ -47,6 +48,12 @@ await on multiple things with timeouts.
     * - :class:`Task`
       - Task object.
 
+    * - :func:`run_coroutine_threadsafe`
+      - Schedule a coroutine from another OS thread.
+
+    * - ``for in`` :func:`as_completed`
+      - Monitor for completeness with a ``for`` loop.
```
Comment (elprans) [inline Doc/library/asyncio-api-index.rst]: completion?

File: Doc/library/asyncio-llapi-index.rst [hunk 2]
```diff
@@ -0,0 +1,511 @@
+.. currentmodule:: asyncio
+
+
+====================
+Low-level APIs Index
```
Comment (elprans) [inline Doc/library/asyncio-llapi-index.rst]: I'd call this "Low-level asyncio APIs" as we don't seem to use the word "index" in other index-like pages.

File: Doc/library/asyncio-llapi-index.rst [hunk 3]
```diff
@@ -0,0 +1,511 @@
...
+====================
+
+This page lists all low-level asyncio APIs.
+
+
+Accessing Event Loop
```
Comment (elprans) [inline Doc/library/asyncio-llapi-index.rst]: Obtaining?

**Review by willingc — APPROVED** [review]
Beautiful! Nice work @1st1

File: Doc/library/asyncio-llapi-index.rst [hunk 4]
```diff
@@ -0,0 +1,511 @@
...
+      - Get the current debug mode.
+
+
+.. rubric:: Scheduling Callbacks
+.. list-table::
+    :widths: 50 50
+    :class: full-width-table
+
+    * - :meth:`loop.call_soon`
+      - Call a callable soon.
```
Comment (elprans) [inline Doc/library/asyncio-llapi-index.rst]: s/Call/Invoke/ to avoid "buttery butter".

File: Doc/library/asyncio-llapi-index.rst [hunk 5]
```diff
@@ -0,0 +1,511 @@
...
+
+    * - :meth:`loop.call_soon_threadsafe`
+      - A thread-safe variant of :meth:`loop.call_soon`.
+
+    * - :meth:`loop.call_later`
+      - Call callback *after* the given time.
```
Comment (elprans) [inline Doc/library/asyncio-llapi-index.rst]: Ditto

File: Doc/library/asyncio-llapi-index.rst [hunk 6]
```diff
@@ -0,0 +1,511 @@
...
+
+    * - :meth:`loop.call_at`
+      - Call callback *at* the given time.
```
Comment (elprans) [inline Doc/library/asyncio-llapi-index.rst]: Ditto

File: Doc/library/asyncio-llapi-index.rst [hunk 7]
```diff
@@ -0,0 +1,511 @@
...
+
+
+.. rubric:: Thread/Process Pool
+.. list-table::
+    :widths: 50 50
+    :class: full-width-table
+
+    * - ``await`` :meth:`loop.run_in_executor`
+      - Run a CPU-bound function in a :mod:`concurrent.futures`
```
Comment (elprans) [inline Doc/library/asyncio-llapi-index.rst]: It doesn't really have to be "CPU-bound", but we can phrase it as a recommendation that "CPU-bound or other blocking functions..."

File: Doc/library/asyncio-llapi-index.rst [hunk 8]
```diff
@@ -0,0 +1,511 @@
...
+.. rubric:: Tasks and Futures
+.. list-table::
+    :widths: 50 50
+    :class: full-width-table
+
+    * - :meth:`loop.create_future`
+      - Create a :class:`Future` object.
+
+    * - :meth:`loop.create_task`
+      - Wrap a coroutine into :class:`Task` and schedule
```
Comment (elprans) [inline Doc/library/asyncio-llapi-index.rst]: I feel like "wrap a coroutine" speaks of an implementation detail and isn't itself useful. "Schedule coroutine as a Task" sounds better to me.

File: Doc/library/asyncio-llapi-index.rst [hunk 9]
```diff
@@ -0,0 +1,511 @@
...
+      - Open a datagram (UDP) connection.
+
+    * - ``await`` :meth:`loop.sendfile`
+      - Send a file over a transport.
+
+    * - ``await`` :meth:`loop.start_tls`
+      - Upgrade an existing connection to TLS.
+
+    * - ``await`` :meth:`loop.connect_read_pipe`
+      - Wrap a read-pipe into a ``(transport, protocol)`` pair.
```
Comment (elprans) [inline Doc/library/asyncio-llapi-index.rst]: "read end of a pipe"

File: Doc/library/asyncio-llapi-index.rst [hunk 10]
```diff
@@ -0,0 +1,511 @@
...
+
+    * - ``await`` :meth:`loop.connect_write_pipe`
+      - Wrap a write-pipe into a ``(transport, protocol)`` pair.
```
Comment (elprans) [inline Doc/library/asyncio-llapi-index.rst]: "write end of a pipe"

File: Doc/library/asyncio-llapi-index.rst [hunk 11]
```diff
@@ -0,0 +1,511 @@
...
+      - Connect the :class:`~socket.socket`.
+
+    * - ``await`` :meth:`loop.sock_accept`
+      - Accept a :class:`~socket.socket` connection.
+
+    * - ``await`` :meth:`loop.sock_sendfile`
+      - Send a file over the :class:`~socket.socket`.
+
+    * - :meth:`loop.add_reader`
+      - Start watching an FD for read availability.
```
Comment (elprans) [inline Doc/library/asyncio-llapi-index.rst]: Let's spell "file descriptor" in full.

File: Doc/library/asyncio-llapi-index.rst [hunk 12]
```diff
@@ -0,0 +1,511 @@
...
+
+    * - :meth:`transport.pause_reading() <ReadTransport.pause_reading>`
+      - Pause receiving.
+
+    * - :meth:`transport.resume_reading() <ReadTransport.resume_reading>`
+      - Resume receiving.
+
+
+Transports that can Send data (TCP and Unix connections,
+write-pipes, etc).  Returned from methods like
```
Comment (elprans) [inline Doc/library/asyncio-llapi-index.rst]: Simply "pipes".  It is clear which end from the context.

File: Doc/library/asyncio-llapi-index.rst [hunk 13]
```diff
@@ -0,0 +1,511 @@
...
+
+    * - :meth:`transport.set_protocol() <BaseTransport.set_protocol>`
+      - Set a new protocol.
+
+    * - :meth:`transport.get_protocol() <BaseTransport.get_protocol>`
+      - Return the current protocol.
+
+
+Transports that can receive data (TCP and Unix connections,
+read-pipes, etc).  Returned from methods like
```
Comment (elprans) [inline Doc/library/asyncio-llapi-index.rst]: Simply "pipes".  It is clear which end from the context.

File: Doc/library/asyncio-llapi-index.rst [hunk 14]
```diff
@@ -0,0 +1,511 @@
...
+
+    * - :meth:`transport.write_eof() <WriteTransport.write_eof>`
+      - Close and send EOF after flushing buffered data.
+
+    * - :meth:`transport.abort() <WriteTransport.abort>`
+      - Close the transport immediately.
+
+    * - :meth:`transport.get_write_buffer_size()
+        <WriteTransport.get_write_buffer_size>`
+      - Return high- and low-water limits for write flow control.
```
Comment (elprans) [inline Doc/library/asyncio-llapi-index.rst]: "High/low water mark" seems to be a better term here.

File: Doc/library/asyncio-llapi-index.rst [hunk 15]
```diff
@@ -0,0 +1,511 @@
...
+
+    * - :meth:`transport.set_write_buffer_limits()
+        <WriteTransport.set_write_buffer_limits>`
+      - Set new high- and low-water limits for write flow control.
```
Comment (elprans) [inline Doc/library/asyncio-llapi-index.rst]: Ditto

File: Doc/library/asyncio-llapi-index.rst [hunk 16]
```diff
@@ -0,0 +1,511 @@
...
+    :class: full-width-table
+
+    * - :meth:`transport.sendto() <DatagramTransport.sendto>`
+      - Send data to the remote peer.
+
+    * - :meth:`transport.abort() <DatagramTransport.abort>`
+      - Close the transport immediately.
+
+
+Low-level transports abstraction over subprocesses.
```
Comment (elprans) [inline Doc/library/asyncio-llapi-index.rst]: I generally prefer singular form for the first part of a compound noun: "transport abstraction".

File: Doc/library/asyncio-llapi-index.rst [hunk 17]
```diff
@@ -0,0 +1,511 @@
...
+    :class: full-width-table
+
+    * - ``callback`` :meth:`connection_made() <BaseProtocol.connection_made>`
+      - Called when a connection is made.
+
+    * - ``callback`` :meth:`connection_lost() <BaseProtocol.connection_lost>`
+      - Called when the connection is lost or closed.
+
+    * - ``callback`` :meth:`pause_writing() <BaseProtocol.pause_writing>`
+      - Called when the transport's buffer goes over the high-water mark.
```
Comment (elprans) [inline Doc/library/asyncio-llapi-index.rst]: s/high-water/high water/

File: Doc/library/asyncio-llapi-index.rst [hunk 18]
```diff
@@ -0,0 +1,511 @@
...
+
+    * - ``callback`` :meth:`resume_writing() <BaseProtocol.resume_writing>`
+      - Called when the transport's buffer drains below the low-water mark.
```
Comment (elprans) [inline Doc/library/asyncio-llapi-index.rst]: Ditto

File: Doc/library/asyncio-llapi-index.rst (hunk 2 above)
Comment (1st1) [reply Doc/library/asyncio-llapi-index.rst]: We use it for high-level cheatsheet too, let's keep this as is (I'll drop the "s" as you suggested though)

**Comment by 1st1** [comment]
Thanks @elprans and @willingc!

**Pull request merged**